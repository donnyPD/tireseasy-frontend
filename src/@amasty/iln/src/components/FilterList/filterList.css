.root {
  padding-bottom: 2rem;
}

.items {
  display: grid;
  gap: 1rem;
  grid-template-columns: 100%;
}

.itemsEmpty {
  composes: items;
}

.itemHidden {
  display: none;
}

.showMoreLessButton {
  margin-top: 1rem;
  text-decoration: underline;
}

.showMoreLessButton:hover {
  text-decoration: none;
}

.swatchList {
  composes: items;
  display: flex;
  flex-wrap: wrap;
  gap: 0;
  margin: -0.5rem;
}

.swatchList:empty {
  margin: 0;
}

.swatchList .item {
  margin: 0.5rem;
}

.fromTo {
  composes: root from '../FromTo/fromTo.css';
  padding: 1rem 0 0 0;
}

.itemsEmpty + .fromTo {
  padding-top: 0;
}
