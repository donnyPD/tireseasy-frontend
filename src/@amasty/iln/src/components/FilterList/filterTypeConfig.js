import PriceSlider from '../PriceSlider';
import Swatch from '../Swatch';
import Rating from '../Rating';
import Category from '../Category';
import defaultClasses from './filterList.css';
import Label from '../Label';
import FromTo from '../FromTo';
import {
  getRatingTitle,
  getStockTitle,
  getRangeTitle,
  getIsNewTitle,
  getOnSaleTitle
} from '../../utils/titles';

export const filterTypeConfig = {
  Slider: {
    listComponent: PriceSlider,
    optionTitle: getRangeTitle
  },
  Ranges: {
    itemComponent: Label,
    optionTitle: getRangeTitle
  },
  Images: {
    itemComponent: Swatch,
    classes: { items: defaultClasses.swatchList }
  },
  'Text Swatches': {
    itemComponent: Swatch,
    classes: { items: defaultClasses.swatchList },
    textOnly: true
  },
  'Images & Labels': {
    itemComponent: Label
  },
  rating: {
    itemComponent: Rating,
    optionTitle: getRatingTitle
  },
  stock: {
    itemComponent: Label,
    optionTitle: getStockTitle
  },
  category_ids: {
    listComponent: Category
  },

  'From-To Only': {
    listComponent: FromTo,
    optionTitle: getRangeTitle
  },

  am_on_sale: {
    itemComponent: Label,
    optionTitle: getOnSaleTitle
  },
  am_is_new: {
    itemComponent: Label,
    optionTitle: getIsNewTitle
  }
};

export default filterTypeConfig;
