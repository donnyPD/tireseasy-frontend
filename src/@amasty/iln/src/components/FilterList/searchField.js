import React from 'react';
import { mergeClasses } from '@magento/venia-ui/lib/classify';
import { func, shape, string } from 'prop-types';
import { useIntl } from 'react-intl';
import defaultClasses from './searchField.css';
import Trigger from '@magento/venia-ui/lib/components/Trigger';
import Icon from '@magento/venia-ui/lib/components/Icon';
import { X as ClearIcon } from 'react-feather';
import TextInput from '@magento/venia-ui/lib/components/TextInput';

const clearIcon = <Icon src={ClearIcon} size={24} />;

const SearchField = props => {
  const { onChange, reset, value, field } = props;

  const { formatMessage } = useIntl();

  const resetButton = value ? (
    <Trigger action={reset}>{clearIcon}</Trigger>
  ) : null;

  const classes = mergeClasses(defaultClasses, props.classes);

  return (
    <div className={classes.root}>
      <TextInput
        placeholder={formatMessage({
          id: 'filterList.search',
          defaultMessage: 'Search'
        })}
        after={resetButton}
        field={field}
        autoComplete="off"
        onValueChange={onChange}
      />
    </div>
  );
};

SearchField.propTypes = {
  onChange: func,
  classes: shape({
    root: string
  })
};

SearchField.defaultProps = {};

export default SearchField;
