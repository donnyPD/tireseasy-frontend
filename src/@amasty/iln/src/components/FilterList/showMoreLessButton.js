import React from 'react';
import { mergeClasses } from '@magento/venia-ui/lib/classify';
import { bool, func, number, shape, string } from 'prop-types';
import defaultClasses from './filterList.css';
import { useIntl } from 'react-intl';

const ShowMoreLessButton = props => {
  const {
    itemCountToShow,
    itemsLength,
    isListExpanded,
    handleListToggle
  } = props;
  const { formatMessage } = useIntl();

  if (!itemCountToShow || itemsLength <= itemCountToShow) {
    return null;
  }

  const count = itemsLength - itemCountToShow;

  const label = isListExpanded
    ? formatMessage({
        id: 'filterList.showLess',
        defaultMessage: 'Show Less'
      })
    : formatMessage(
        {
          id: 'amIln.showMore',
          defaultMessage: 'Show ({count}) More'
        },
        { count }
      );

  const classes = mergeClasses(defaultClasses, props.classes);

  return (
    <button
      onClick={handleListToggle}
      className={classes.showMoreLessButton}
      type="button"
    >
      {label}
    </button>
  );
};

ShowMoreLessButton.propTypes = {
  itemCountToShow: number,
  itemsLength: number,
  isListExpanded: bool,
  handleListToggle: func,
  classes: shape({
    root: string
  })
};

export default ShowMoreLessButton;
