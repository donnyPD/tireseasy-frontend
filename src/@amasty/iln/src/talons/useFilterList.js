import { useCallback, useMemo, useState } from 'react';
import { useFieldApi } from 'informed';

export const useFilterList = (props = {}) => {
  const {
    group,
    filterApi,
    onApply,
    filterBlockSettings,
    items,
    filterState
  } = props;

  const [isListExpanded, setExpanded] = useState(false);
  const searchFieldName = `search_${group}`;

  const [searchValue, setSearchValue] = useState('');
  const { reset } = useFieldApi(searchFieldName);

  const {
    is_multiselect: isMultiSelect,
    is_show_search_box,
    number_unfolded_options: itemCountToShow,
    show_product_quantities,
    add_from_to_widget: isVisibleFromTo,
    limit_options_show_search_box
  } = filterBlockSettings;

  const visibleItems = useMemo(() => {
    let list = items;

    if (isVisibleFromTo) {
      list = items.filter(({ custom }) => !custom);
    }

    if (searchValue) {
      list = items.filter(({ title }) =>
        title.toLowerCase().includes(searchValue.toLowerCase())
      );
    }
    return new Set(list.map(({ value }) => value));
  }, [items, searchValue, isVisibleFromTo]);

  const handleListToggle = useCallback(() => {
    setExpanded(value => !value);
  }, [setExpanded]);

  const { toggleItem, changeItem } = filterApi;

  const handleSelectItem = useCallback(
    item => {
      const isSelected = filterState && filterState.has(item);
      const duplicatedItem = filterState && Array.from(filterState).find(({ value }) => value === item.value);

      if (!isMultiSelect && !isSelected) {
        changeItem({ group, item });
      } else if (!duplicatedItem || isSelected) {
        toggleItem({ group, item });
      }

      if (typeof onApply === 'function') {
        onApply(group, item);
      }
    },
    [group, isMultiSelect, toggleItem, onApply, changeItem, filterState]
  );

  const changeSearch = useCallback(value => setSearchValue(value), [
    setSearchValue
  ]);

  const resetSearch = useCallback(() => reset(), [reset]);

  return {
    handleListToggle,
    isListExpanded,
    itemCountToShow,
    visibleItems,
    searchProps: {
      field: searchFieldName,
      onChange: changeSearch,
      value: searchValue,
      reset: resetSearch,
      isShowSearchBox:
        is_show_search_box &&
        (!limit_options_show_search_box ||
          limit_options_show_search_box < items.length)
    },

    itemProps: {
      showProductQuantities: !!show_product_quantities,
      isMultiSelect,
      handleSelectItem
    },
    isVisibleFromTo
  };
};
