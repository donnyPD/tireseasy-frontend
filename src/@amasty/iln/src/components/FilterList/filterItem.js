import React, { useMemo } from 'react';
import { func, number, oneOfType, shape, string } from 'prop-types';
import setValidator from '@magento/peregrine/lib/validators/set';
import Label from '../Label';
import { useFilterItem } from '../../talons/useFilterItem';

const FilterItem = props => {
  const { filterState, item, handleSelectItem, itemComponent, ...rest } = props;

  const talonProps = useFilterItem({
    filterState,
    item,
    handleSelectItem
  });
  const { isSelected, tileItem, handleClick } = talonProps;

  const Component = useMemo(() => itemComponent || Label, [itemComponent]);

  return (
    <Component
      isSelected={isSelected}
      item={tileItem}
      onClick={handleClick}
      {...rest}
    />
  );
};

FilterItem.propTypes = {
  filterState: setValidator,
  item: shape({
    title: string.isRequired,
    value: oneOfType([number, string]).isRequired
  }).isRequired,
  onChange: func
};

export default FilterItem;
