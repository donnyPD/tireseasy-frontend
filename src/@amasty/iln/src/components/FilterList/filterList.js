import React, { useMemo } from 'react';
import { array, shape, string, func } from 'prop-types';
import { mergeClasses } from '@magento/venia-ui/lib/classify';
import defaultClasses from './filterList.css';
import { useFilterList } from '../../talons/useFilterList';
import setValidator from '@magento/peregrine/lib/validators/set';
import FilterItem from './filterItem';
import SearchField from './searchField';
import ShowMoreLessButton from './showMoreLessButton';
import FromTo from '../FromTo';

const labels = new WeakMap();

const FilterList = props => {
  const { filterState, items, group, ...rest } = props;

  const {
    isListExpanded,
    handleListToggle,
    itemCountToShow,
    itemProps,
    searchProps,
    visibleItems,
    isVisibleFromTo
  } = useFilterList(props);

  const classes = mergeClasses(defaultClasses, props.classes);

  const itemElements = useMemo(
    () =>
      items.map((item, index) => {
        const { title, value } = item;
        const key = `item-${group}-${value}`;
        const itemClass =
          visibleItems.has(value) &&
          (isListExpanded || !itemCountToShow || index < itemCountToShow)
            ? classes.item
            : classes.itemHidden;

        const element = (
          <li key={key} className={itemClass}>
            <FilterItem
              filterState={filterState}
              item={item}
              {...itemProps}
              {...rest}
            />
          </li>
        );

        labels.set(element, title.toUpperCase() || '');

        return element;
      }),
    [
      classes,
      filterState,
      items,
      isListExpanded,
      itemCountToShow,
      itemProps,
      group,
      visibleItems,
      rest
    ]
  );

  const { isShowSearchBox } = searchProps;

  const searchBox = isShowSearchBox ? <SearchField {...searchProps} /> : null;
  const fromToWidget = isVisibleFromTo ? (
    <FromTo {...props} classes={{ root: classes.fromTo }} />
  ) : null;
  const itemsClass = visibleItems.size ? classes.items : classes.itemsEmpty;

  return (
    <div className={classes.root}>
      {searchBox}
      <ul className={itemsClass}>{itemElements}</ul>
      <ShowMoreLessButton
        itemCountToShow={itemCountToShow}
        itemsLength={items.length}
        isListExpanded={isListExpanded}
        handleListToggle={handleListToggle}
      />

      {fromToWidget}
    </div>
  );
};

FilterList.defaultProps = {
  onApply: null,
  isExpanded: false
};

FilterList.propTypes = {
  classes: shape({
    item: string,
    items: string
  }),
  filterApi: shape({}),
  filterState: setValidator,
  group: string,
  items: array,
  onApply: func
};

export default FilterList;
