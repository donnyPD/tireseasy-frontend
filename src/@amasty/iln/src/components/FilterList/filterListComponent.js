import React from 'react';
import FilterList from './filterList';
import { filterTypeConfig } from './filterTypeConfig';

const FilterListComponent = props => {
  const { filterBlockSettings, filterState, items, ...restProps } = props;
  const { display_mode_label } = filterBlockSettings;

  const filterType = filterTypeConfig[display_mode_label] || {};
  const { listComponent, ...rest } = filterType;

  const ListComponent = listComponent || FilterList;

  const componentProps = {
    filterBlockSettings,
    ...restProps,
    ...rest
  };

  return (
    <ListComponent
      {...componentProps}
      items={items}
      filterState={filterState}
    />
  );
};

export default FilterListComponent;
