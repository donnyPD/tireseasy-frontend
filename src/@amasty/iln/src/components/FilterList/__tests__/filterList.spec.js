import React from 'react';
import { createTestInstance } from '@magento/peregrine';
import FilterList from '../filterList';
import { useFilterList } from '../../../talons/useFilterList';

jest.mock('../filterItem', () => 'FilterItem');
jest.mock('../searchField', () => 'SearchField');
jest.mock('../showMoreLessButton', () => 'ShowMoreLessButton');

jest.mock('../../../talons/useFilterList', () => {
  const useFilterList = jest.fn(() => {});

  return { useFilterList };
});

const props = {
  items: [
    {
      title: 'test1',
      value: '1'
    },
    {
      title: 'test2',
      value: '2'
    }
  ]
};

test('render correct list', () => {
  useFilterList.mockReturnValue({
    isListExpanded: true,
    itemCountToShow: 5,
    searchProps: {},
    visibleItems: new Set()
  });

  const tree = createTestInstance(<FilterList {...props} />);
  expect(tree.toJSON()).toMatchSnapshot();
});

test('render correct list with search', () => {
  useFilterList.mockReturnValue({
    isListExpanded: true,
    itemCountToShow: 5,
    searchProps: {
      isShowSearchBox: true
    },
    visibleItems: new Set()
  });

  const tree = createTestInstance(<FilterList {...props} />);
  expect(tree.toJSON()).toMatchSnapshot();
});
