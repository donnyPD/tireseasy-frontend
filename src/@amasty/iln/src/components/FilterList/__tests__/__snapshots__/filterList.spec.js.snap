// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correct list 1`] = `
<div>
  <ul>
    <li>
      <FilterItem
        isExpanded={false}
        item={
          Object {
            "title": "test1",
            "value": "1",
          }
        }
        onApply={null}
      />
    </li>
    <li>
      <FilterItem
        isExpanded={false}
        item={
          Object {
            "title": "test2",
            "value": "2",
          }
        }
        onApply={null}
      />
    </li>
  </ul>
  <ShowMoreLessButton
    isListExpanded={true}
    itemCountToShow={5}
    itemsLength={2}
  />
</div>
`;

exports[`render correct list with search 1`] = `
<div>
  <SearchField
    isShowSearchBox={true}
  />
  <ul>
    <li>
      <FilterItem
        isExpanded={false}
        item={
          Object {
            "title": "test1",
            "value": "1",
          }
        }
        onApply={null}
      />
    </li>
    <li>
      <FilterItem
        isExpanded={false}
        item={
          Object {
            "title": "test2",
            "value": "2",
          }
        }
        onApply={null}
      />
    </li>
  </ul>
  <ShowMoreLessButton
    isListExpanded={true}
    itemCountToShow={5}
    itemsLength={2}
  />
</div>
`;
